eventsources/sources/bitbucketserver/start.go

```diff
@@ -0,0 +1,274 @@
+
+	apiResponse, err := router.bitbucketClient.DefaultApi.FindWebhooks(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to list existing hooks to check for duplicates for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	hooks, err := bitbucketv1.GetWebhooksResponse(apiResponse)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert the list of webhooks for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	var existingHook bitbucketv1.Webhook
+	isAlreadyExists := false
+
+	for _, hook := range hooks {
+		if hook.Url == formattedURL {
+			existingHook = hook
+			isAlreadyExists = true
+		}
+	}
+
+	logger.Info("retrieving the webhook secret...")
+	webhookSecret, err := router.getWebhookSecret(bitbucketserverEventSource.WebhookSecret)
+	if err != nil {
+		return errors.Errorf("failed to get bitbucketserver webhook secret. err: %+v", err)
+	}
+
+	newHook := bitbucketv1.Webhook{
+		Name:          "Argo Events",
+		Url:           formattedURL,
+		Active:        true,
+		Events:        bitbucketserverEventSource.Events,
+		Configuration: bitbucketv1.WebhookConfiguration{Secret: webhookSecret.secret},
+	}
+
+	localVarPostBody, err := json.Marshal(newHook)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marhal new webhook to JSON")
+	}
+
+	if !isAlreadyExists {
+		apiResponse, err = router.bitbucketClient.DefaultApi.CreateWebhook(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, localVarPostBody, []string{"application/json"})
+
+		if err != nil {
+			return errors.Errorf("failed to add webhook. err: %+v", err)
+		}
+
+	} else {
+		logger.Info("webhook already exists, updating it...")
+
+		apiResponse, err = router.bitbucketClient.DefaultApi.UpdateWebhook(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, int32(existingHook.ID), localVarPostBody, []string{"application/json"})
+
+		if err != nil {
+			return errors.Errorf("failed to update webhook. err: %+v", err)
+		}
+	}
+
+	logger.Infof("created webhook apiResponse: %+v", apiResponse)
+
+	var createdHook *bitbucketv1.Webhook
+	err = mapstructure.Decode(apiResponse.Values, &createdHook)
+
+	logger.Infof("created webhook: %+v", createdHook)
+
+	if err != nil {
+		return errors.Errorf("failed to convert API response to Webhook struct. err: %+v", err)
+	}
+
+	router.hook = createdHook
+	logger.With("hook-id", createdHook.ID).Info("hook succesfully registered")
+	return nil
+}
+
+// PostInactivate performs operations after the route is inactivated
+func (router *Router) PostInactivate() error {
+	bitbucketserverEventSource := router.bitbucketserverEventSource
+	route := router.route
+
+	if bitbucketserverEventSource.DeleteHookOnFinish {
```

Comment:
Race conditions will happen because of this, see https://github.com/argoproj/argo-events/issues/1177. Check github event source for the workaround.

Reply:
Please note that I copied the boilerplate for this EventSource from the Gitlab EventSource, so that probably means Gitlab is suffering from the same race conditions. The comments in the Github EventSource make it seem that it's a Github-specific issue ("Github can not handle race conditions well"), but in fact all EventSources that depend on external resources being created/destroyed suffer from the same issue (currently gitlab and awssns, and I believe the Stripe EventSource will create a webhook for every replica).

IMHO, adding the workaround is a bit too much of a kludge and the logic for managing external EventSource dependencies should move elsewhere. The current workaround causes the webhooks to be deleted and recreated for every replica and only works correctly when the number of replicas in a RollingUpdate stays the same (or is higher), otherwise the deletes "win" the race.

Reply:
> Please note that I copied the boilerplate for this EventSource from the Gitlab EventSource, so that probably means Gitlab is suffering from the same race conditions. The comments in the Github EventSource make it seem that it's a Github-specific issue ("Github can not handle race conditions well"), but in fact all EventSources that depend on external resources being created/destroyed suffer from the same issue (currently gitlab and awssns, and I believe the Stripe EventSource will create a webhook for every replica).
> 

This is true. I never tried Gitlab and others, not sure if they handle the race well.

> IMHO, adding the workaround is a bit too much of a kludge and the logic for managing external EventSource dependencies should move elsewhere. The current workaround causes the webhooks to be deleted and recreated for every replica and only works correctly when the number of replicas in a RollingUpdate stays the same (or is higher), otherwise the deletes "win" the race.

I don't like the webhook creation being handled by the eventsource, and wanted to deprecate it, https://github.com/argoproj/argo-events/issues/1112, but there were objections. The workaround works, because it keeps checking and creating.

Reply:
> I don't like the webhook creation being handled by the eventsource, and wanted to deprecate it, #1112, but there were objections. The workaround works, because it keeps checking and creating.

I really like the fact that the webhooks are set-up automatically. It's also shown in one of the [videos](https://www.youtube.com/watch?v=XNXJtxkUKeY) linked from the Argo-Events homepage.

I've now added the same race condition handling code to the bitbucketserver eventsource and it seems to work fine.